### PR TITLE
peergos: init at 0.14.1

### LINKED ISSUE #78362: package request: Peergos
**Project description**
An end-to-end encrypted, peer-to-peer file storage, sharing and communication network.
It's written in Java and uses IPFS. A bit similar to Nextcloud. 

**Metadata**

* homepage URL: https://peergos.org/
* source URL: https://github.com/Peergos/Peergos
* license: AGPL3
* platforms: linux (JVM)

### PR DESCRIPTION
**Pull request by raspher**
## Description of changes

Peergos is an end-to-end encrypted, peer-to-peer file storage, sharing and communication network. https://peergos.org/

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [X] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by mattpolzin — COMMENTED** [review]
You'll need to add your new package to `pkgs/top-level/all-packages.nix` (you can grep for other packages living in the `applications/networking/p2p/` subdirectory to see how they are included).

I'm pretty sure you'll also need to make sure each commit of yours is formatted correctly; your first commit was, but the second one that fixes formatting was not. In this case, I think I'd suggest rebasing and squashing your second commit into the first one anyway.

**Comment by donovanglover** [comment]
Welcome to nixpkgs! Note that the new [`pkgs/by-name`](https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#name-based-package-directories) format should be used where possible

**Comment by mattpolzin** [comment]
Ah, and while I'm already making a suggestion to squash one of your commits I should also mention that the addition to the maintainer-list file should be made as a separate commit with the message "maintainers: add raspher"

There are some fiddly best practices to worry about but you're not too far off!

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (dotlambda) [inline pkgs/applications/networking/p2p/peergos/default.nix:20]:
```suggestion
    makeWrapper ${lib.getExe jre} $out/bin/peergos \
      --chdir $out/share/peergos/ \
      --add-flags "-Djava.library.path=$out/share/peergos/native-lib/libtweetnacl.so -jar $out/share/java/peergos.jar"
```
Why is chdir needed?

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (dotlambda) [inline pkgs/applications/networking/p2p/peergos/default.nix:22]: Is this supposed to be a test?

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (dotlambda) [inline pkgs/applications/networking/p2p/peergos/default.nix:23]:
```suggestion
  '';
```

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (dotlambda) [inline pkgs/applications/networking/p2p/peergos/default.nix:10]: Put this in a `let`.

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (dotlambda) [inline pkgs/applications/networking/p2p/peergos/default.nix:18]:
```suggestion
    install -D ${peergos} $out/share/java/peergos.jar
```

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (dotlambda) [inline pkgs/applications/networking/p2p/peergos/default.nix:28]:
```suggestion
    license = licenses.gpl3Only;
```
as far as I can tell

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (raspher) [reply pkgs/applications/networking/p2p/peergos/default.nix:20]: It's because running `peergos` creates folder _native-lib/libtweetnacl.so_ in current `pwd`

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (raspher) [reply pkgs/applications/networking/p2p/peergos/default.nix:22]: should create `native-lib/libtweetnacl.so` 

anyway when running `$out/bin/peergos` it does not recognize this file and fallbacks to pure Java crypto library - see https://github.com/Peergos/Peergos/issues/849

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (raspher) [reply pkgs/applications/networking/p2p/peergos/default.nix:28]: it's agpl on github

**Comment by raspher** [comment]
- moved default.nix to `pkgs/by-name/pe/peergos/package.nix`
- applied some suggestions

Now i'm waiting for feedback as not everything is clear to me, maybe i do something strange in `pakcage.nix`.

Also forgot to mention linked issue: #78362

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (dotlambda) [reply pkgs/applications/networking/p2p/peergos/default.nix]: You should remove it, see below

Comment (dotlambda) [reply pkgs/applications/networking/p2p/peergos/default.nix]: Can we pass an environment variable or do to change the location of the file?

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (dotlambda) [reply pkgs/applications/networking/p2p/peergos/default.nix]: Please add a comment explaining that the file is generated.

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (dotlambda) [reply pkgs/applications/networking/p2p/peergos/default.nix]: Link?

**Comment by dotlambda** [comment]
Please switch the order of the commits.

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (raspher) [reply pkgs/applications/networking/p2p/peergos/default.nix]: [peergos](https://github.com/Peergos/Peergos)... Sorry, now i see [web-ui](https://github.com/Peergos/web-ui) has GPL3, what license should i use in this case? They're both used in this package.

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (raspher) [reply pkgs/applications/networking/p2p/peergos/default.nix]: I don't think so, it seems to be a bug. I was trying to do that by setting `-Djava.library.path` but it has no effect.

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (dotlambda) [reply pkgs/applications/networking/p2p/peergos/default.nix]: You csn use list containing both and add a comment explaining it.

File: pkgs/applications/networking/p2p/peergos/default.nix
Comment (dotlambda) [reply pkgs/applications/networking/p2p/peergos/default.nix]: In that case please open an upstream issue/file a PR.

File: pkgs/by-name/pe/peergos/package.nix
Comment (mattpolzin) [inline pkgs/by-name/pe/peergos/package.nix:24]: This is all pretty minor feeling IMHO, but: 
- You probably don't want to inherit `peergos` as a property of the derivation (no need for that in order to use the sources below).
- You could move `pname` and `meta` back into the body of the derivation just leaving `version` and `peergos` in your `let` block since `peergos` is the only one that doesn't need to be a derivation property and `version` is the only variable the `fetchurl` call depends on.

**Review by mattpolzin — COMMENTED** [review]
I'd like to leave approval for someone who has experience packaging other Java projects (that's not me) but you've addressed all of my other suggestions and my remaining two points might be optional. Thanks!

**Comment by mattpolzin** [comment]
This did not build for me on `x86_64-darwin`, though.
```
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "configurePhase" }
Running phase: configurePhase
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
Running phase: buildPhase
no Makefile or custom buildPhase, doing nothing
@nix { "action": "setPhase", "phase": "installPhase" }
Running phase: installPhase
java.lang.IllegalStateException: /var/empty/.peergos/peergos.0.log.lck
        at peergos.server.util.Logging.init(Logging.java:77)
        at peergos.server.util.Logging.init(Logging.java:38)
        at peergos.server.Command.main(Command.java:74)
        at peergos.server.Main.main(Main.java:906)
Caused by: java.nio.file.NoSuchFileException: /var/empty/.peergos/peergos.0.log.lck
        at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
        at java.base/sun.nio.fs.UnixFileSystemProvider.newFileChannel(UnixFileSystemProvider.java:181)
        at java.base/java.nio.channels.FileChannel.open(FileChannel.java:304)
        at java.base/java.nio.channels.FileChannel.open(FileChannel.java:363)
        at java.logging/java.util.logging.FileHandler.openFiles(FileHandler.java:512)
        at java.logging/java.util.logging.FileHandler.<init>(FileHandler.java:451)
        at java.logging/java.util.logging.FileHandler.<init>(FileHandler.java:409)
        at peergos.server.util.Logging.init(Logging.java:60)
        ... 3 more
```

**Comment by mattpolzin** [comment]
Oddly enough, I get roughly the same result when running in an `x86_64-linux` VM. Not sure why the difference, but I am generally able to test linux builds this way.
```
building '/nix/store/rf62567p9r7hrip2m39b6hkn7z65phbj-peergos-0.14.1.drv'...
Running phase: patchPhase
Running phase: updateAutotoolsGnuConfigScriptsPhase
Running phase: configurePhase
no configure script, doing nothing
Running phase: buildPhase
no Makefile or custom buildPhase, doing nothing
Running phase: installPhase
Couldn't load native crypto library at /nix/store/dg9zs7rkhi801wyaibm31vbn2w30c8m0-peergos-0.14.1/share/peergos/native-lib/libtweetnacl.so, using pure Java version...
To use the native linux-x86-64 crypto implementation use option -Djava.library.path=native-lib
java.lang.IllegalStateException: /var/empty/.peergos/peergos.0.log.lck
        at peergos.server.util.Logging.init(Logging.java:77)
        at peergos.server.util.Logging.init(Logging.java:38)
        at peergos.server.Command.main(Command.java:74)
        at peergos.server.Main.main(Main.java:906)
Caused by: java.nio.file.NoSuchFileException: /var/empty/.peergos/peergos.0.log.lck
        at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
        at java.base/sun.nio.fs.UnixFileSystemProvider.newFileChannel(UnixFileSystemProvider.java:181)
        at java.base/java.nio.channels.FileChannel.open(FileChannel.java:304)
        at java.base/java.nio.channels.FileChannel.open(FileChannel.java:363)
        at java.logging/java.util.logging.FileHandler.openFiles(FileHandler.java:512)
        at java.logging/java.util.logging.FileHandler.<init>(FileHandler.java:451)
        at java.logging/java.util.logging.FileHandler.<init>(FileHandler.java:409)
        at peergos.server.util.Logging.init(Logging.java:60)
        ... 3 more
error: builder for '/nix/store/rf62567p9r7hrip2m39b6hkn7z65phbj-peergos-0.14.1.drv' failed with exit code 255
```

**Comment by raspher** [comment]
```bash
$ ./result/bin/peergos 
Jan 23, 2024 5:59:00 PM peergos.server.util.Logging init
INFO: Logging to /home/raspher/.peergos/peergos.0.log
Run with -help to show options
```
On local build using `nix-shell -p` and `nix-build -E 'with import <nixpkgs> {}; callPackage ./package.nix {}' -v` it was fine. Maybe it's related to generating .so files (comments in `package.nix` may explain). I previously had generated those files by just running binary which wants to log to `~/.peergos/...`

Can you retry build and run this?

**Comment by mattpolzin** [comment]
Built and launched executable on `x86_64-linux` and `x86_64-darwin` as of latest commit. I'd recommend removing the `platforms` restriction now.

**Comment by raspher** [comment]
Platform restrictions are removed. Is there something I can do to get this merged into the master? I've seen some packages waiting for merge for  a long time and I don't understand why. Also a lot of pull requests are waiting for merge.

**Comment by mattpolzin** [comment]
There are a number of reasons pull requests to nixpkgs can take a while to merge, but in my experience it can often be difficult to get the right person's eyes on a PR which can leave some PRs sitting for a while if not forever. It's complicated to manage volunteer reviewers on a scale as large as nixpkgs. Patience is definitely important, but a bit of gracious nudging can also help your cause. 

You may have already gotten fortunate to have eyes on this PR from @dotlambda in that respect. Other things you can try to get more attention include pinging the other two assigned reviewers for this PR (sometimes assignment to review a PR lands in a different notification category than at-mentions) and mentioning this PR in a "review needed" thread on either Discord (unofficial) or Discourse. I don't have the links handy but there are nixos/nixpkgs groups for both of those services. The review threads/channels are pre-existing, so don't create a new thread, just a new post/comment.

Hope that helps.

**Comment by raspher** [comment]
Result of `nixpkgs-review pr 282470` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>peergos</li>
  </ul>
</details>

File: pkgs/by-name/pe/peergos/package.nix
Comment (dotlambda) [inline pkgs/by-name/pe/peergos/package.nix:35]: This doesn't make sense to me. What would `cd /nix/store/..-peergos/share/peergos` help with if that directory isn't writable?

File: pkgs/by-name/pe/peergos/package.nix
Comment (dotlambda) [inline pkgs/by-name/pe/peergos/package.nix:17]: Can we put this at the bottom as is usually done?

File: pkgs/by-name/pe/peergos/package.nix
Comment (dotlambda) [inline pkgs/by-name/pe/peergos/package.nix:23]:
```suggestion
    maintainers = with maintainers; [ raspher ];
    sourceProvenance = with sourceTypes; [ binaryBytecode ];
```

File: pkgs/by-name/pe/peergos/package.nix
Comment (dotlambda) [inline pkgs/by-name/pe/peergos/package.nix:26]:
```suggestion
  dontUnpack = true;
```
I guess we also want `dontBuild`.

File: pkgs/by-name/pe/peergos/package.nix
Comment (dotlambda) [inline pkgs/by-name/pe/peergos/package.nix:31]:
```suggestion
    runHook preInstall

    install -D ${peergos} $out/share/java/peergos.jar
```

File: pkgs/by-name/pe/peergos/package.nix
Comment (dotlambda) [inline pkgs/by-name/pe/peergos/package.nix:36]:
```suggestion

    runHook postInstall
  '';
```

File: pkgs/by-name/pe/peergos/package.nix
Comment (raspher) [reply pkgs/by-name/pe/peergos/package.nix:35]: It won't create `native-libs` folder, please look at mentioned issues

Comment (dotlambda) [reply pkgs/by-name/pe/peergos/package.nix:35]: I did. So if the directory is not writable then `native-libs` won't be created, which is what we want?
In that case why not `--chdir $out/nowhere`?
I still don't think we should use `--chdir` because then relative paths passed on the command line won't be interpreted correctly.

Comment (raspher) [reply pkgs/by-name/pe/peergos/package.nix:35]: What do you think I should do here? Wait for a fix for that? We can of course

Comment (dotlambda) [reply pkgs/by-name/pe/peergos/package.nix:35]: The upstream behaviour is to create native-libs in cwd, so we should just go with that behavior until they fix it. Or is the problem that native-libs doesn't work on NixOS? Btw is native-libs downloaded from the internet or compiled on the machine?

Comment (raspher) [reply pkgs/by-name/pe/peergos/package.nix:35]: looks like they have the `.so` file on repo and it's included into the .jar file - their [build.xml](https://github.com/Peergos/Peergos/blob/12ec589e841fd51233bff15d4d3eb27761efca15/build.xml#L81). It seems to be same file loaded into jar on every architecture and unpacked to current directory to use it. Altrough i'm not java expert and may be wrong. You're right, i'll remove the chdir and let unpack this file where they want.

**Comment by raspher** [comment]
Done

File: pkgs/by-name/pe/peergos/package.nix
Comment (dotlambda) [inline pkgs/by-name/pe/peergos/package.nix:32]: Can we put that at the top?

File: pkgs/by-name/pe/peergos/package.nix
Comment (dotlambda) [inline pkgs/by-name/pe/peergos/package.nix:16]:
```suggestion
  pname = "peergos";
  inherit version;
```

**Review by dotlambda — APPROVED** [review]

**Comment by raspher** [comment]
Thanks, finally this package looks beautiful and i've learned a lot

**Comment by raspher** [comment]
Ping @piegamesde @Janik-Haag

**Pull request merged**
